format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/SplitMultilineIterator.java

```diff
@@ -0,0 +1,125 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.format.delimited.input;
+
+import com.google.common.collect.AbstractIterator;
+import org.apache.hadoop.mapreduce.RecordReader;
+
+import java.io.IOException;
+
+/**
+ * Iterator that provides the splits in the delimited string based on the delimiter. The delimiter
+ * should not contain any quotes. The spliterator will behave like this: 1. if there is no quote, it
+ * will behave same as {@link String#split(String)} 2. if there are quotes in the string, the method
+ * will find pairs of quotes, content within each pair of quotes will not get split even if there
+ * is delimiter in that. For example, if string is a."b.c"."d.e.f" and delimiter is '.', it will get
+ * split into [a, b.c, d.e.f]. if string is "val1.val2", then it will not get split since the '.'
+ * is within a pair of quotes. If data is spread across multiple lines and quote is started within first line, it
+ * will iterate over next lines until closing quote is there.If the delimited string contains odd number of quotes,
+ * which mean the quotes are not closed, an exception will be thrown. The quote within the value will always be trimmed.
+ */
+public class SplitMultilineIterator extends AbstractIterator<String> {
+  private static final char QUOTE_CHAR = '\"';
+  private final RecordReader recordReader;
+  private final String delimiter;
+  private final String lineSeparator = "\n";
+  private String delimitedString;
+  private int index;
+  private boolean endingWithDelimiter = false;
+
+  public SplitMultilineIterator(String delimitedString, String delimiter, RecordReader recordReader) {
+    this.delimitedString = delimitedString;
+    this.delimiter = delimiter;
+    this.recordReader = recordReader;
+    index = 0;
+  }
+
+  @Override
+  protected String computeNext() {
+    // Corner case when the delimiter is in the end of the row
+    if (endingWithDelimiter) {
+      endingWithDelimiter = false;
+      return "";
+    }
+
+    boolean isWithinQuotes = false;
+
+    if (index == delimitedString.length()) {
+      return endOfData();
+    }
+
+    StringBuilder split = new StringBuilder();
+    while (index < delimitedString.length()) {
+      char cur = delimitedString.charAt(index);
+      if (cur == QUOTE_CHAR) {
+        isWithinQuotes = !isWithinQuotes;
+        index++;
+        appendNextLineToDelimitedString(isWithinQuotes);
+        continue;
+      }
+
+      // if the length is not enough for the delimiter or it's not a delimiter, just add it to split
+      if (index + delimiter.length() > delimitedString.length() ||
+        !delimitedString.startsWith(delimiter, index)) {
+        split.append(cur);
+        index++;
+        appendNextLineToDelimitedString(isWithinQuotes);
```

Comment:
strings are immutable, so appending like this is actually creating a new string and can be relatively costly. It also means the string grows ever longer and uses more memory. 

It would be more efficient to append a newline to the split, set delimitedString to the record reader's next line, and reset the index to 0.

Reply:
done.